bionic/deriver.py

```diff
@@ -443,63 +393,201 @@ def tasks(self):
+            # computed in main process vs subprocess when entity is not persisted.
+            # NOTE 2: Right now, non-persisted entities include simple lookup values
+            # which we should not be really sending using IPC. We should read/write
+            # a tmp file for this instead to use protocol for serialization instead of
+            # using cloudpickle.
+            state.complete(task_key_logger)
+            self._transition_entry(entry, "COMPLETE")
+        else:
+            # NOTE 1: Logging support for multiple processes not done yet.
+            new_state_for_subprocess = state.strip_state_for_subprocess()
+            future = self._bootstrap.process_executor.submit(
+                new_state_for_subprocess.complete, task_key_logger
+            )
+            entry.future = future
+            self._transition_entry(entry, "IN_PROGRESS")
+
+        entry.visited = True
+
+    def _get_or_create_entry_for_state(self, state):
+        task_key = state.task.keys[0]
+        if task_key not in self._entries_by_task_key:
+            self._entries_by_task_key[task_key] = TaskRunnerEntry(state=state)
+        return self._entries_by_task_key[task_key]
+
+    def _has_more_ready_entries(self):
+        # While there are no ready states but has any in-progress ones, let's wait
+        # for in-progress entries to finish till we find a ready entry or exhaust all
+        # in-progress entries.
+        while len(self._ready_entries) == 0 and len(self._in_progress_entries) != 0:
+            self._wait_on_in_progress_entries()
+        return len(self._ready_entries) != 0
+
+    def _next_ready_entry(self):
+        assert len(self._ready_entries) != 0
+        return self._ready_entries.pop()
+
+    def _wait_on_in_progress_entries(self):
+        "Waits on any in-progress entry to finish."
+        futures = [
+            entry.future
+            for entry in self._in_progress_entries.values()
+            if entry.future is not None
         ]
-        blockage = TaskBlockage(blocked_state, blocking_tks)
-        for blocking_tk in blocking_tks:
-            self._blockage_lists_by_blocking_tk[blocking_tk].append(blockage)
+        finished_futures, _ = wait(futures, return_when=FIRST_COMPLETED)
+        for finished_future in finished_futures:
+            task_key = finished_future.result()
+            entry = self._entries_by_task_key[task_key]
+            self._transition_entry(entry, "COMPLETE")
 
-    def get_unblocked_by(self, completed_state):
+    def _transition_entry(self, entry, transition_type):
         """
-        Records the fact that a task state is complete, and yields all task states
-        that are newly unblocked.
+        Handles entry transition and updates cached entries.
+
+        Parameters
+        ----------
+
+        entry: TaskRunnerEntry
+            Entry to transition.
+
+        transition_type: String with allowed values in [
```

Comment:
I was thinking of it as more of an action to apply rather than the next stage. The issue with keeping the next stage was that I had to keep it in sync with the task state status. For example, we don't have a blocked stage on runner entry. However, the code is much simplified now and might be easier to attach a stage enum to keep track of an entry's progress, like how you expanded on in your [other comment](https://github.com/square/bionic/pull/120#discussion_r419537877).